aptos-move/aptos-vm/src/aptos_vm.rs

```diff
@@ -462,23 +462,29 @@ impl AptosVM {
             let init_function = session.load_function(&module.self_id(), init_func_name, &[]);
             // it is ok to not have init_module function
             // init_module function should be (1) private and (2) has no return value
-            if init_function.is_ok() {
-                if verify_module_init_function(module).is_ok() {
-                    let args: Vec<Vec<u8>> = senders
-                        .iter()
-                        .map(|s| MoveValue::Signer(*s).simple_serialize().unwrap())
-                        .collect();
-                    session.execute_function_bypass_visibility(
-                        &module.self_id(),
-                        init_func_name,
-                        vec![],
-                        args,
-                        gas_meter,
-                    )?;
-                } else {
-                    return Err(PartialVMError::new(StatusCode::CONSTRAINT_NOT_SATISFIED)
-                        .finish(Location::Undefined));
-                }
+            match init_function {
+                Ok(_) => if verify_module_init_function(module).is_ok() {
+                        let args: Vec<Vec<u8>> = senders
+                            .iter()
+                            .map(|s| MoveValue::Signer(*s).simple_serialize().unwrap())
+                            .collect();
+                        session.execute_function_bypass_visibility(
+                            &module.self_id(),
+                            init_func_name,
+                            vec![],
+                            args,
+                            gas_meter,
+                        )?;
+                    } else {
+                        return Err(PartialVMError::new(StatusCode::CONSTRAINT_NOT_SATISFIED)
+                            .finish(Location::Undefined));
+                    },
+                Err(err) => {
+                    if err.major_status() != StatusCode::FUNCTION_RESOLUTION_FAILURE {
```

Comment:
I'm a bit confused of what case could be silently ignored before. Because this is just the loader shouldn't any failure here be ignorable? It doesn't seem to be an execution failure, because here the init function isn't yet executed.

Reply:
I guess I may have had too many tests there and it may have got lost.
If you look at `state_v0` and `use_state_error` in the PR, it should be clear. And also something you can try on your machine if you patch this PR.
I have updated the test to point out to what to run to get the error (I should have done it earlier).
The bottom line is that if we have an `init_module` that is compiled against code introduced in an update (and the init is in the update obviously), then, though verification of the bundle is good and correct, when trying to load the init, it gets linked against the "old" module, which does not contain the new function. Thus a "verifier" error.
Hope that explains a bit better

Reply:
Looks like we can't detect an transitive resolution failure so `fun init_module() { some_future::fun() }` will be still silent.

Reply:
let me see if I understand. 
The point here is that once we leave the module where the `init_module` is, we are out of the upgrade/update and so we will not get any verification/linker error any longer because we are using existing ("old" so to speak) modules.
So this only catches first level "errors" (or at least the example is one case of such first level error).
Please explain the scenario if I missed it, as there may be even more edge cases, and I'd love to know them.
That was my difficulty in coming to terms with this and the reason for the internal conversation we had.
Whether we "fix"/patch it or not is the question here, and I personally understand if there are concerns/doubts/perplexity around this PR and what we are really fixing.

Overall though I think @movekevin question can be phrased as simply as "should we report an error when we get it or ignore it and publish anyway"? 
And we could ask ourselves that question irrespective of catching every error and being consistent.
Should we error when we know there is an error and the init will not run?

Reply:
or thinking some more, maybe the point in your example is that if the init contains a reference to a new module introduced in the upgrade we also have an issue!
Let me write a test about it and add it to the PR as that could help.
Thanks!

Reply:
I'm not sure why the long and winded response, everything looks fine, just saying that the heuristic might not complete, which is totally fine.

Reply:
got it, I thought you were pointing out another test case! 
(and that actually made me think about stuff, so thanks!)